# Data Analysis with NumPy

![numpy.jpg](numpy.jpg)

## Importing Numpy and Pandas

In [2]:
import numpy as np
import pandas as pd

# Importing Data

In [4]:
data = 'data/forestfires.csv'

df = pd.read_csv(data)

In [5]:
df.head()

,X,Y,month,day,FFMC,DMC,DC,ISI,temp,RH,wind,rain,area
0,7,5,mar,fri,86.2,26.2,94.3,5.1,8.2,51,6.7,0.0,0.0
1,7,4,oct,tue,90.6,35.4,669.1,6.7,18.0,33,0.9,0.0,0.0
2,7,4,oct,sat,90.6,43.7,686.9,6.7,14.6,33,1.3,0.0,0.0
3,8,6,mar,fri,91.7,33.3,77.5,9.0,8.3,97,4.0,0.2,0.0
4,8,6,mar,sun,89.3,51.3,102.2,9.6,11.4,99,1.8,0.0,0.0


## Dataset Description

Attribute Information -

The dataset contains 517 instances and 13 attributes. The attribute information is as follows:-

* X - x-axis spatial coordinate within the Montesinho park map: 1 to 9
* Y - y-axis spatial coordinate within the Montesinho park map: 2 to 9
* month - month of the year: 'jan' to 'dec'
* day - day of the week: 'mon' to 'sun'
* FFMC - FFMC index from the FWI system: 18.7 to 96.20
* DMC - DMC index from the FWI system: 1.1 to 291.3
* DC - DC index from the FWI system: 7.9 to 860.6
* ISI - ISI index from the FWI system: 0.0 to 56.10
* temp - temperature in Celsius degrees: 2.2 to 33.30
* RH - relative humidity in %: 15.0 to 100
* wind - wind speed in km/h: 0.40 to 9.40
* rain - outside rain in mm/m2 : 0.0 to 6.4
* area - the burned area of the forest (in ha): 0.00 to 1090.84

In [7]:
fires = df.iloc[:5, 8:]
print(fires)

   temp  RH  wind  rain  area
0   8.2  51   6.7   0.0   0.0
1  18.0  33   0.9   0.0   0.0
2  14.6  33   1.3   0.0   0.0
3   8.3  97   4.0   0.2   0.0
4  11.4  99   1.8   0.0   0.0


# NumPy ndarray object

numpy.array(object, dtype=None, copy=True, order=None, subok=False, ndmin=0)

The above constructor takes the following parameters:
* object - Any object exposing the array interface method returns an array, or any (nested) sequence.
* dtype - Desired data type of array, optional.
* copy - (Optional) - By default (true), the object is copied.
* order - C (row major) or F (column major) or A (any) (default).
* subok - By default, returned array forced to be a base class array. If true, sub-classes passed through.
* ndmin - Specifies minimum dimensions of resultant array.

In [8]:
# one-dimensional array
a = np.array([1,2,3])
print(a)

[1 2 3]


In [11]:
n1 = np.array(fires[:1], dtype=np.float32)
print(n1)

[[ 8.2 51.   6.7  0.   0. ]]


In [12]:
# two-dimensional array
b = np.array([
    [1,2],
    [3,4]
])

print(b)

[[1 2]
 [3 4]]


In [13]:
n2 = np.array(fires[:2], dtype=int)

print(n2)

[[ 8 51  6  0  0]
 [18 33  0  0  0]]


In [14]:
# there-dimensional array
c = np.array([
    [1,2],
    [3,4],
    [5,6]
])

print(c)

[[1 2]
 [3 4]
 [5 6]]


In [15]:
n3 = np.array(fires[:3], dtype=np.int64)

print(n3)

[[ 8 51  6  0  0]
 [18 33  0  0  0]
 [14 33  1  0  0]]


In [21]:
# minimum dimensions

d = np.array([1,2,3,4,5], ndmin=2)
print(d)

[[1 2 3 4 5]]


In [22]:
# dtype parameter

e = np.array([1,2,3,4,5], dtype=complex)

print(e)

[1.+0.j 2.+0.j 3.+0.j 4.+0.j 5.+0.j]


## NumPy data types

In [24]:
dt = np.dtype(np.int32)

print(dt)

int32


In [25]:
# create structured data type
dt = np.dtype([('age', np.int8)])
print(dt)

[('age', 'i1')]


In [26]:
# now apply it to ndarray object
dt = np.dtype([('age', np.int8)])

arr1 = np.array([(10,), (20,), (30,)], dtype=dt)
print(arr1)

[(10,) (20,) (30,)]


In [27]:
# file name can be used to access content of age column
dt = np.dtype([('age', np.int8)])

arr2 = np.array([(10,), (20,), (30,)], dtype=dt)
print(arr2['age'])

[10 20 30]


In [28]:
student = np.dtype([('name', 'S20'), ('age', 'i1'), ('marks', 'f4')])

print(student)

[('name', 'S20'), ('age', 'i1'), ('marks', '<f4')]


In [29]:
student = np.dtype([('name', 'S20'), ('age', 'i1'), ('marks', 'f4')])
student_data = np.array([('Krasi', 23, 50), ('Vlado', 24, 55)], dtype=student)
print(student_data)

[(b'Krasi', 23, 50.) (b'Vlado', 24, 55.)]


## NumPy array attributes

### ndarray.shape

In [31]:
x1 = np.array([
    [1,2,3],
    [4,5,6]
])

print(x1.shape)

(2, 3)


In [32]:
print(n1.shape)

(1, 5)


In [33]:
print(n2.shape)

(2, 5)


In [36]:
x2 = np.array([
    [1,2,3],
    [4,5,6]
])

x2.shape = (3,2)

print(x2)

[[1 2]
 [3 4]
 [5 6]]


### ndarray.size

In [37]:
x1 = np.array([
    [1,2,3],
    [4,5,6]
])

print(x1.size)

6


In [38]:
print(n1.size)

5


In [39]:
print(n2.size)

10


In [40]:
print(n3.size)

15


### reshape function

In [41]:
x1 = np.array([
    [1,2,3],
    [4,5,6]
])

x1_reshaped = x1.reshape(3,2)
print(x1_reshaped)

[[1 2]
 [3 4]
 [5 6]]


In [42]:
n3 = n3.reshape(5,3)
print(n3)

[[ 8 51  6]
 [ 0  0 18]
 [33  0  0]
 [ 0 14 33]
 [ 1  0  0]]


### ndarray.ndim

In [43]:
x3 = np.arange(24)

print(x3)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]


In [44]:
print(x3.ndim)

1


In [45]:
x4 = x3.reshape(2,4,3)
print(x4)

[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]
  [ 9 10 11]]

 [[12 13 14]
  [15 16 17]
  [18 19 20]
  [21 22 23]]]


In [46]:
print(x4.ndim)

3


In [47]:
print(n1.ndim)

2


### ndarray.itemsize

In [62]:
x5 = np.array([1,2,3,4,5], dtype=np.int64)
print(x5)

[1 2 3 4 5]


In [63]:
print(x5.itemsize)

8


In [59]:
x6 = np.array([1,2,3,4,5], dtype=np.float16)

print(x6)

[1. 2. 3. 4. 5.]


In [60]:
print(x6.itemsize)

2


### ndarray.dtype

In [64]:
print(x5.dtype.name)

int64


In [65]:
print(x6.dtype.name)

float16


In [69]:
print(n1.dtype.name)

float32


In [70]:
print(n2.dtype.name)

int64


### type of array

In [71]:
type(x1)

numpy.ndarray

In [73]:
type(x2)

numpy.ndarray

In [74]:
type(n1)

numpy.ndarray

## NumPy array creation

A new ndarray object can be constructed by any of the following array creation routines. There are several ways to create arrays. These are listed below:
* array function
* empty function
* zeros function
* ones function
* arange function
* linspace function

### array function

In [76]:
x7 = np.array([1,2,3])
x7

array([1, 2, 3])

In [77]:
x7.dtype

dtype('int64')

In [78]:
list1 = [1,2,3,4]

a1 = np.array(list1)
print(a1.dtype)

int64


In [88]:
dict1 = {
    "a": 1,
    "b": 2,
    "c": 3,
}

a2 = np.array(dict1, dtype=np.object_)
print(a2)

{'a': 1, 'b': 2, 'c': 3}


In [89]:
print(a2.dtype)

object


### empty function

In [102]:
x9 = np.empty([10,3], dtype=int)
print(x9)

[[      2766105950440                 188                   0]
 [                  0   32088490942398464 7235419174270214779]
 [7291434879409994274 7005459366441661494 7074419657749572919]
 [7292235105020568422 3832898879714834534 8243121639892462626]
 [2459029315949257070 7308604897285731189 8319104481851616290]
 [3540427691124485993 3270006080094168675 7234527220238475878]
 [4047998727391161653 7004334587605508402 7215364926925660472]
 [8461736183933006945 7454421801564400748 2322206376936961119]
 [7310597164893758754 8391162085809418847 8318823019759807522]
 [3828658067276197737   12947851028084014    3377699728457840]]


In [104]:
z1 = np.empty([1,10], dtype=int)
print(z1)

[[-1  0  0  0  0  0  0  0  0  0]]


### zeros function

In [107]:
z3 = np.zeros(5, dtype=int)
print(z3)

[0 0 0 0 0]


In [111]:
z2 = np.zeros([10,2,3], dtype=int)
print(z2)

[[[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]]]


In [112]:
z4 = np.zeros([2,2], dtype=np.int_)
print(z4)

[[0 0]
 [0 0]]


### ones function

In [113]:
np.ones(5, dtype=int)

array([1, 1, 1, 1, 1])

In [115]:
x10 = np.ones([2,10,10], dtype=int)
print(x10)

[[[1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]]

 [[1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]
  [1 1 1 1 1 1 1 1 1 1]]]


In [117]:
x11 = np.ones([2,10,10], dtype=np.float64)
print(x11)

[[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]

 [[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]]


### arange function

In [119]:
b1 = np.arange(10,20,2)
print(b1)

[10 12 14 16 18]


In [122]:
b2 = np.arange(10,20,0.5)
print(b2)

[10.  10.5 11.  11.5 12.  12.5 13.  13.5 14.  14.5 15.  15.5 16.  16.5
 17.  17.5 18.  18.5 19.  19.5]


### linspace function

In [126]:
c1 = np.linspace(start=2.0, stop=4.0, num=50)
print(c1)

[2.         2.04081633 2.08163265 2.12244898 2.16326531 2.20408163
 2.24489796 2.28571429 2.32653061 2.36734694 2.40816327 2.44897959
 2.48979592 2.53061224 2.57142857 2.6122449  2.65306122 2.69387755
 2.73469388 2.7755102  2.81632653 2.85714286 2.89795918 2.93877551
 2.97959184 3.02040816 3.06122449 3.10204082 3.14285714 3.18367347
 3.2244898  3.26530612 3.30612245 3.34693878 3.3877551  3.42857143
 3.46938776 3.51020408 3.55102041 3.59183673 3.63265306 3.67346939
 3.71428571 3.75510204 3.79591837 3.83673469 3.87755102 3.91836735
 3.95918367 4.        ]


In [127]:
c2 = np.linspace(start=2.0, stop=4.0, num=5, endpoint=False)
print(c2)

[2.  2.4 2.8 3.2 3.6]


## NumPy array from existing data